### PR TITLE
Add trigger and reactions check func

### PR DESCRIPTION
**Pull request by romainpanno**
MAJOR

* **Checklist before requesting a review**
- [x] The commits messages follows our guidelines
- [x] I have performed a self-review of my code
- [x] I have documented my code
- [ ] Do we need to implement unit tests?

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of changes does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **Describe your changes**
* Adding trigger check and reaction action when trigger is done

* **Issue ticket number and link**

## Summary by CodeRabbit

- **New Features**
  - Implemented a task scheduling system that triggers actions at regular intervals.
  - Introduced interfaces for action and reaction functions within the task scheduler.
  - Added new methods for retrieving action and reaction data from the database.

- **Enhancements**
  - Updated the `ActionData` and `ReactionData` models with new properties and renamed existing ones for better clarity.
  - Improved encryption service to generate a new initialization vector for each encryption.

- **Bug Fixes**
  - Ensured that OAuth tokens are now stored as encrypted and have mandatory initialization vectors for enhanced security.
  
- **Documentation**
  - Added comments and documentation for new functions and models.

- **Refactor**
  - Reorganized import statements to improve codebase consistency.
  - Renamed the OAuth class for standardization.
  
- **Chores**
  - Checked for the presence of essential environment variables during app initialization.

### CODE REVIEW

File: back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * @description Spotify reactions services
+ */
+namespace SpotifyReactions {
+    /**
+     * Add a song to a playlist
+     * @param actionData - The action data
+     * @param reactionData - The reaction data
+     * @returns {Promise<boolean>} - The result of the reaction
+     */
+    export const reactionSpotifyAddToPlaylist = async (actionData: JSON, reactionData: JSON): Promise<boolean> => {
```
[the remark this answers is not in the document]
Comment (romainpanno) [reply back/src/services/servicesApps/reactions/spotify.reactions.servicesApp.ts:16]: we will do it later, or check inside the function the type in order to do error handling inside the function himself, is it good ?

File: back/src/index.ts [hunk 2]
```diff
@@ -27,6 +29,17 @@ app.get('/', (req, res) => {
...
+/**
+ * Execute each minute the checkTriggers function, which checks if a trigger is activated
+ */
+setInterval(async () => {
+    try {
+        await TaskScheduler.checkTriggers();
+    } catch (error) {
+        console.error('Erreur lors de la vérification des triggers:', error);
+    }
+}, INTERVAL); // 60000 ms = 1 minute
```
[the remark this answers is not in the document]
Comment (romainpanno) [reply back/src/index.ts:41]: how to do that ?

File: back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+namespace SpotifyTriggers {
+    /**
+     * Check if a new song has been saved
+     * @param ownerId - The owner id of the trigger
+     * @param oauthId - The oauth id of the trigger
+     * @param data - The data of the trigger
+     * @returns {Promise<boolean>} - The result of the trigger
+     */
+    export const checkSpotifyNewSavedSong = async (ownerId: number, oauthId: number, data: JSON): Promise<boolean> => {
+        console.log(ownerId);
```
[the remark this answers is not in the document]
Comment (romainpanno) [reply back/src/services/servicesApps/triggers/spotify.triggers.servicesApp.ts:18]: This code will be removed it's only for test, so no logger with different log levels need to be implemented at this time

File: back/src/services/encryption.service.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+if (!process.env.CRYPTO_SECRET) {
+    throw new Error('No secret key defined');
+}
+
+/* Constants */
+const ALGORITHM: string = process.env.CRYPTO_ALGO;
+const SECRET_KEY: string = process.env.CRYPTO_SECRET; // Must be 256 bytes (32 characters)
+const IV: Buffer = crypto.randomBytes(16); // Initialization vector
```
[the remark this answers is not in the document]
Comment (romainpanno) [reply back/src/services/encryption.service.ts:20]: don't understand, how to do that ?

File: back/src/models/action.model.ts [hunk 5]
```diff
@@ -1,42 +1,42 @@
...
-        },
-        name: {
-          type: DataTypes.STRING,
-          unique: true,
-          allowNull: false,
-        },
-        args: {
-          type: DataTypes.JSON,
-          allowNull: true,
-
```
Comment (Vivelis) [inline back/src/models/action.model.ts:27]: should not be an empty line

File: back/src/models/action.model.ts [hunk 6]
```diff
@@ -1,42 +1,42 @@
...
+                name: {
+                    type: DataTypes.STRING,
+                    unique: true,
+                    allowNull: false,
+                },
+                args: {
+                    type: DataTypes.JSON,
+                    allowNull: true,
+
+                },
```
Comment (Vivelis) [inline back/src/models/action.model.ts:28]: allowNull: true,
        
        },
```suggestion
                    allowNull: true,
                },
```

File: back/src/models/oauth.model.ts [hunk 7]
```diff
@@ -1,42 +1,47 @@
...
-          allowNull: true,
-        },
-      },
-      {
-        sequelize,
-        tableName: 'oauth',
-      }
-    );
-  }
+    public id!: number;
```
Comment (Vivelis) [inline back/src/models/oauth.model.ts:7]: Should be no empty line here
```suggestion
class OAuth extends Model {
    public id!: number;
```

File: back/src/models/reaction.model.ts [hunk 8]
```diff
@@ -1,42 +1,42 @@
...
+                name: {
+                    type: DataTypes.STRING,
+                    unique: true,
+                    allowNull: false,
+                },
+                args: {
+                    type: DataTypes.JSON,
+                    allowNull: true,
+
+                },
```
Comment (Vivelis) [inline back/src/models/reaction.model.ts:28]: There should be no empty line here
```suggestion
                    allowNull: true,
                },
```

File: back/src/models/reactionData.model.ts [hunk 9]
```diff
@@ -1,41 +1,46 @@
...
+                },
+                ownerId: {
+                    type: DataTypes.INTEGER,
+                    allowNull: false,
+                },
+                data: {
+                    type: DataTypes.JSON,
+                    allowNull: true,
+
+                },
```
Comment (Vivelis) [inline back/src/models/reactionData.model.ts:28]: There should be no empty line here
```suggestion
                    allowNull: true,
                },
```

File: back/src/models/service.model.ts [hunk 10]
```diff
@@ -1,42 +1,42 @@
...
-          allowNull: true,
-        },
-      },
-      {
-        sequelize,
-        tableName: 'services',
-      }
-    );
-  }
+    public id!: number;
```
Comment (Vivelis) [inline back/src/models/service.model.ts:7]: There should be no empty line here.
```suggestion
class Service extends Model {
    public id!: number;
```

**Review by Vivelis — APPROVED** [review]

**Review by badidu09 — APPROVED** [review]
Nice code

**Pull request merged**